### DISCUSSION TITLE
Framework for custom "plugins"

### DISCUSSION THREAD

**Discussion by bilogic** [arendst/Tasmota#12011] [Q&A] [answered]
Hi,

I have a "problem", but it isn't a bug with Tasmota, thought I make this clear first.
I was also concerned it might get lost in Discord, that's the reason why I'm creating this issue here.

### PROBLEM DESCRIPTION
I have custom Tasmota code and find it a challenge to rebase my changes on to the latest version of Tasmota.

2 of my main customizations are:

1. Custom RF `xdrv_17_rcswitch.ino` to support both receiving RF315 and RF433 concurrently
2. Control of normal servos to open and close lids/covers

The reason I need my own code is because they can be quite hard to configure in the official Tasmota. And in the case of servos, I feel, the complexity of controlling a servo should not be "programmed" through configuration.

I'm aware it is always a triage to decide which features to cram into the ESP, so, having a framework to support custom code should eliminate this "fight". To someone like me, Tasmota's value lies in its Wifi, MQTT and reporting mechanisms which interfaces very well with software like HomeAssistant and NodeRed, something which I would like to leverage on.

So, I hope to suggest a way to make integrating customizations much more easier:

1. My idea is to pre-allocate a number of blank `xdrv_*`, `xsns_*` etc for customizations
2. In the main branch, they will be left blank with maybe just some example code
3. This way, once there is a new Tasmota release, we can just replace these files and automate our custom compilation

I came up with the idea only last night and have not scrutinized through every detail.
Open to hearing any opinions on this. Thank you.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): _____
- [ ] Tasmota binary firmware version number used: _____
  - [ ] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by arendst** [comment, marked as answer]
I suppose you've read the documentation regarding API https://tasmota.github.io/docs/API/ and https://tasmota.github.io/docs/Sensor-API/

The idea was that users could easily add their drivers or sensors using an ID from 90 to 97.

So to add seamlessly you just need to select an unused driver or sensor id in that range.

**Reply by bilogic** [reply]
Are there predefined `UserSelectablePins` for ID 90 to 97?
I would like let the user define which pin the custom driver was added to.

**Comment by bilogic** [comment]
No, I have not. Let me have a read. Thank you.

**Comment by sfromis** [comment]
With ESP32, there is also the new feature of the [Berry](https://tasmota.github.io/docs/Berry/) scripting language, which may be a more approachable programming environment than, say, creating a driver in C++. I found it quite straightforward to implement a driver for an unsupported I2C sensor.

**Reply by bilogic** [reply]
I see, so that's what Berry is about. However, I have a change in mindset recently, preferring to write 0 to minimal code #lowcode.

I managed to use a library to get my servo working in a few lines in C++. So, first question to my mind is, will Berry require me to go deeper into servo technicals (which I hope to avoid), e.g. PWM, wavelength etc?

Further more, if this works the way I think it does, my plan is to leverage on other well built libraries + convert all my custom code into Tasmota extensions. (I suspect Berry might not be suitable here)

Either way, I will have a quick look at Berry. Thank you.

**Reply by sfromis** [reply]
The need to go into technicalities when creating a driver is basically similar for Berry and C++, but as a scripting language Berry may relieve you of the C++ requirement to be very strict about a lot of details. Of course, for time-critical code, Berry may not work out. Obviously, Berry is an entirely different direction than leveraging existing libraries.

**Comment by ozett** [comment]
i am also struggling to get my minimal-driver code into tasmota
is there more help on a SAMPLE or FRAMEWORK for drivers than the developer readme?

a working bare minimum code sample would be nice to have.

i only found [tasmota-custom-integration](https://github.com/phpcoder/tasmota-custom-integration)

but there seems to be serious steps for making a customize build with platform.io beforehand to make, - to have a clean codebase.

anybody maybe has some websites/links for more on this?

**Comment by ozett** [comment]
assuming that the discussion is the right place to ask:
how do i bind my driver to the GPIO dropdown?
What is the mechanism for that? i barely find some instructions for a good start.
seems some automation runs behind it. but not obvious what to configure for that. some coding or directives?

<kbd> ![image](https://user-images.githubusercontent.com/5937811/216252566-07aa6c0e-1d90-4ba8-a3a9-98992ecc4698.png)

**Reply by barbudor** [reply]
I would suggest that you start by looking at a given driver which makes uses of a few pins (if your driver needs specific GPIO by opposition to an I2C driver for example)

There are some rules to add GPIOS that have to be declared in 2 or 3 places + all translatable texts in all langage files

I would suggest that you open a dedicated discussion giving more details on what is the purpose of your driver, what GPIOs you would need to add and we could start from here to drive you in the right direction

Note also that on ESP32 you can write quite good drivers in Berry language which requires no integration in Tasmota code or PlatformIO

**Comment by ozett** [comment]
seems all only partly documented,
and more than 1 step to go, but i found something

https://github.com/phpcoder/tasmota-custom-integration/issues/3#issuecomment-1413921584
<kbd>![image](https://user-images.githubusercontent.com/5937811/216371860-fcd53238-6ed0-482a-9705-d3de60b175ad.png)

there is long info over the assembler code [for handling uint_32](https://tasmota.github.io/docs/Sensor-API/#first-example),
(and alwas in the issues and discussion the link to this developer section)
and i would wish that there were a straighter documention for gettin the dropdowns

there must have been many people doing this over the years, but only small documents about this architecture.
hope that not only theo arends knows how all this is tied together

will definitely dive into berry. looks promising.
hope that i can flash a single led two times for 30ms in a row with a single pause of 10ms, all in one cylce
as far as i see (as an rookie-coder) in the callback-timings, that can only easy be done within a driver 
but maybe berry can ...

**Reply by sfromis** [reply]
Beware that such very old threads can be somewhat outdated when it comes to details of how things work now.

**Reply by ozett** [reply]
i am experimenting with this in visualstudiocode with the latest checkout 12.2.x
every error on compile i come to the discussions and issues to find hopefully some small pieces of docu how it works..

thanks for havin an eye on this old topic 👍

**Comment by Jason2866** [comment]
@ozett  You can see in this open PR how to add entrys. https://github.com/arendst/Tasmota/pull/17857

**Reply by barbudor** [reply]
@ozett yes documentation for developers could be better
May be it's on purpose to filter the most determined 🤣🤣

Feel free to improve the doc with your findings afterwards

Do not hesitate to make your fork public and ask for advice or comments

**Reply by ozett** [reply]
Really appealing is the access to nearly-realtime with scripting, means (the ESP32) BERRY scripting. 
All compiled in by default. Great!

seems no need to compile a (updated) driver as new firmware. (as long as ESP32 is used).
Great. i will try to experiment with this ..

https://tasmota.github.io/docs/Berry/#creating-a-tasmota-driver
